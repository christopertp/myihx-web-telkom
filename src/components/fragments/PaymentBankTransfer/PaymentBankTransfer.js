import React, { useContext, useEffect, useRef, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';
import { AppContext } from '../../../contexts';
import Bullet from '../../elements/Bullet/Bullet';
import Button from '../../elements/Button';
import Card from '../../elements/Card';
import Heading from '../../elements/Heading';
import IconButton from '../../elements/IconButton';
import useTimer from '../../../hooks/useTimer';
import WhiteCardPlain from '../../layouts/WhiteCardPlain';
import { thousand } from '../../../utils/format';
import styles from './styles.scoped.css';

export default function PaymentBankTransfer() {
  const history = useHistory();
  const { bank, code, detail, method } = useParams();

  useEffect(() => {
    detail && history.replace(`/payment-${code}/${method}`);
  }, []);

  return (
    <>
      {(b => {
        if (b) {
          return <BankDetail />;
        } else {
          return <ListBank />;
        }
      })(bank)}
      <Notif />
    </>
  );
}

export function ListBank() {
  const dispatch = useDispatch();
  const { method } = useParams();
  const { amount, transactionId, bankInfo, fetchBankTransfer } = useSelector(s => s.payment);

  const subtitle = 'Pilih Bank yang dituju';

  if (!bankInfo) {
    return null;
  }

  return (
    <WhiteCardPlain subtitle={subtitle} title="Pilih Bank">
      <section className={styles.root}>
        {
          bankInfo.map((v, i) => (
            <section key={i}
              onClick={() => {
                dispatch(fetchBankTransfer(method, amount, v.codeBank, transactionId, v));
              }}>
              <strong>{v.name}</strong>
              <img alt="chevron red" src="/assets/ic_chevron_red.svg" />
            </section>
          ))
        }
      </section>
    </WhiteCardPlain>
  );
}

export function BankDetail() {
  const dispatch = useDispatch();
  const accountNumberRef = useRef(null);
  const { amount, paymentCode, detailBank, loading,
    transactionId, setMessageAction } = useSelector(s => s.payment);

  const copyToClipboard = e => {
    accountNumberRef.current.select();
    document.execCommand('copy');
    e.target.focus();
    dispatch(setMessageAction('Copied!'));
  };

  return (
    <section className={styles.cardroot}>
      <IconButton className={styles.back} name="close_red" to="/" />
      <header className={styles.logo}>
        {loading.b ? <span className="loading" /> : <img alt="logo bank" src={detailBank.icon} />}
      </header>
      <section className={styles.account}>
        <Heading>Account Number</Heading>
        {loading.b ? <span className="loading" /> : <input readOnly ref={accountNumberRef} type="text" value={paymentCode} />}
        <Button onClick={copyToClipboard}>COPY</Button>
      </section>
      <section className={styles.amount}>
        <Heading>Transfer Amount</Heading>
        {loading.b ? <span className="loading" /> : <strong>Rp {thousand(amount)}</strong>}
        <Heading>Payment Code</Heading>
        {loading.b ? <span className="loading" /> : <strong>{transactionId}</strong>}
      </section>
      <Timer />
      <section className={styles.accordion}>
        <Heading>Lihat cara pembayaran</Heading>
        {loading.b ? <span className="loading" /> : <HowToPay />}
      </section>
      <Button to="/">TUTUP</Button>
    </section>
  );
}

export function HowToPay() {
  const { detailBank } = useSelector(s => s.payment);
  const [idxOpen, setIdxOpen] = useState(-1);

  const height = (item, idx) => {
    let row = 0;
    let rowHeight = 5;
    if (item.howToPay) {
      row = item.howToPay.length;
      rowHeight = ((row * 16) / 3.8 + 2);
    }
    return idx === idxOpen ? `${rowHeight}rem` : '3.5rem';
  };

  const onClickOpen = idx => () => {
    setIdxOpen(idx === idxOpen ? -1 : idx);
  };

  return (
    detailBank && detailBank.details.map((i, idx) => (
      <Card className={styles['channel-card']} key={idx} style={{ height: height(i, idx) }}>
        <header className={styles.headerCard} onClick={onClickOpen(idx)}>
          <h6>{i.name}</h6>
          <IconButton className={idx === idxOpen ? styles.open : ''} name="chevron_red_down" onClick={onClickOpen(idx)} />
        </header>
        <ul>
          {i.howToPay && i.howToPay.map((v, vdx) =>
            (<li key={vdx}>
              <Bullet connect={vdx < i.howToPay.length - 1} />
              <h5>{v.header}</h5>
              <p style={{ paddingBottom: `${!v.description ? '2.1rem' : '0.7rem'}` }}>{v.description}</p>
            </li>)
          )}
        </ul>
      </Card>
    ))
  );
}

export function Notif() {
  const { message } = useSelector(s => s.payment);
  const { setNotif } = useContext(AppContext);

  useEffect(() => {
    setNotif(message);
  }, [message]);

  return null;
}

export function Timer() {
  const { deadline } = useSelector(s => s.payment);
  const currentTime = new Date();
  const expireTime = new Date(deadline);
  const timeInSeconds = (expireTime - currentTime) / (1000 * 60) * 60;

  const { time } = useTimer(Math.floor(timeInSeconds));
  const [hour, min, sec] = time;
  return (
    <div className={styles.timer}>
      <img alt="time" src="/assets/ic_selection_time.svg" />
      <span>Complete payment by</span>
      <span>{hour}:{min}:{sec}</span>
    </div>
  );
}
