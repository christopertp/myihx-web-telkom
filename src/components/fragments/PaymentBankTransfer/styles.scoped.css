.accordion {
  margin-bottom: 1rem;
}

.account {
  box-shadow: 0 0 1rem rgba(0, 0, 1, 0.15);
  border-radius: 0.5rem 0.5rem 0 0;
  background: linear-gradient(0deg, #f3f3f3 0%, #fff 65%);
  display: grid;
  grid-template-columns: auto 1fr;
  margin-top: 1rem;
  padding: 1rem;
}

.account > input, .account > input:focus{
  border: none;
  background-color: transparent;
  font-size: 1rem;
  font-weight: 900;
  outline: none;
}

.account > button {
  border-radius: 0.25rem;
  grid-row: 1/3;
  grid-column: 2;
  margin: auto 0 auto auto;
}

.account > h4 {
  font-size: 1.3rem;
  padding-top: 0.5rem;
}

.account > small {
  font-size: 1rem;
  padding-bottom: 1rem;
}

.amount {
  box-shadow: 0 0.5rem 1rem rgba(0, 0, 1, 0.15);
  border-radius: 0 0 0.5rem 0.5rem;
  display: grid;
  padding: 1rem;
}

.amount > strong:first-of-type {
  margin-bottom: 1rem;
}

.back {
  height: 3rem;
  padding: 0.5rem 0.5rem 0.5rem 0;
}

.cardroot {
  background-color: #fff;
  border-radius: 0.5rem;
  box-shadow: 0 0.5rem 1rem rgba(0, 0, 1, 0.15);
  color: #333435;
  margin: auto;
  max-width: 45rem;
  padding: 1.5rem;
}

.channel-card {
  height: 3.5rem;
  margin-top: 1rem;;
  overflow-y: hidden;
  transition: height 0.25s ease-in-out;
}

.channel-card > header {
  align-items: center;
  cursor: pointer;
  display: flex;
  margin: -1rem;
  padding: 1rem;
}

.channel-card > header > h6 {
  font-size: 1rem;
  font-weight: normal;
  margin-right: auto;
}

.channel-card > header > p {
  color: #b2b4b5;
  margin-right: 0.5rem;
  font-size: 0.75rem;
}

.channel-card > header > button.open {
  transform: rotateZ(180deg);
  transition: transform 0.25s ease-in-out;
}

.channel-card > ul {
  margin-top: 1rem;
}

.channel-card > ul > li {
  display: grid;
  grid-template-columns: 2rem auto;
  grid-template-rows: auto auto;
  max-width: 36rem;
}
.channel-card > ul > li > div {
  grid-row: 1/3;
  margin-right: auto;
  padding-top: 0.5rem;
}
.channel-card > ul > li > h5 {
  font-size: 1rem;
}
.channel-card > ul > li > p {
  color: #66686a;
  font-size: 0.8rem;
  padding-bottom: 0.75rem;
}

.loading {
  border-radius: 0.5rem;
  display: block;
  height: 1.5rem;
  margin-bottom: 0.5rem;
}

.logo > img {
  width: 15rem;
}

.root {
  padding: 1.5rem 0;
}

.root > section {
  align-items: center;
  border-bottom: 1px solid #e5e6e6;
  cursor: pointer;
  display: flex;
  justify-content: space-between;
  padding: 1rem 0;
}

.root > section > strong {
  color: #333435;
  font-size: 1rem;
  font-weight: 800;
}

.root > section > img { height: 1.3rem; }

.timer {
  box-shadow: 0 0.5rem 1rem rgba(0, 0, 1, 0.15);
  border-radius: 0.5rem;
  display: grid;
  grid-template-columns: auto 1fr;
  grid-template-rows: auto auto auto;
  margin: 1.5rem 0;
  padding: 1.25rem 1rem;
}

.timer > img {
  grid-column: 1;
  grid-row: 1/4;
  margin: auto 1rem auto 0;
  width: 3.5rem;
}

.timer > span:first-of-type, .timer > span:last-of-type, .summary-required > h3, .confirm {
  color: #66686a;
  font-size: 0.9rem;
  font-weight: 400;
  letter-spacing: 0.0625rem;
  line-height: 1.6;
}

.timer > span:first-of-type, .summary-required > h3, .confirm { text-transform: uppercase; }

.timer > span:nth-of-type(2) {
  color: var(--primary-color);
  font-size: 1.75rem;
  font-weight: 800;
  line-height: 1.2;
}

@media only screen and (max-width: 48em) {

  .account > button {
    font-size: 0.7rem;
    padding: 0.5rem 1rem;
  }

  .amount {
    background-color: #fff;
  }

  .cardroot {
    background-color: transparent;
    box-shadow: none;
    margin-bottom: 1.75rem;
    min-width: 100%;
    padding: 0;
  }

  .channel-card > ul > li > h5 {
    font-size: 1em;
    margin-top: 0.3rem;
  }
  .channel-card > ul > li > p {
    font-size: 0.7rem;
  }

  .logo > img {
    width: 10rem;
  }

  .root {
    margin: 0 -1rem -1rem;
    padding: 3rem 1rem;
  }
  .root > a {
    border-bottom: 1px solid #ededed;
    padding: 0.75rem 0.5rem 0.75rem 0;
  }
  .root > a > span { font-size: 1rem; }
  .root > a > img { height: 1.2rem; }

  .timer > img, .summary-required > button > img:first-child {
    width: 3rem;
  }

  .timer > span:first-of-type, .timer > span:last-of-type, .summary-required > h3, .summary-required > button > p {
    font-size: 0.75rem;
  }
    
  .timer > span:nth-of-type(2) {
    font-size: 1.25rem;
  }
}
