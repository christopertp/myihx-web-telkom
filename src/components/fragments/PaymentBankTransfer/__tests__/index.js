import React, { useState, useContext, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import ShallowRenderer from 'react-test-renderer/shallow';
import useTimer from '../../../../hooks/useTimer';
import PaymentBankTransfer, { ListBank, BankDetail, HowToPay, Notif, Timer } from '../PaymentBankTransfer';

useSelector.mockImplementation(fn => {
  fn({});
  return {
    amount: 392.50,
    code: '',
    bankInfo: [{
      codeBank: '',
      details: [{
        name: 'MOBILE',
        howToPay: [{
          header: 'dolor ipsum',
          description: 'dolor ipsum',
        }],
      }],
    }],
    fetchBankTransfer: jest.fn(),
    paymentCode: '396525',
    detailBank: {
      icon: '',
      details: [{
        name: 'MOBILE',
        howToPay: [{
          header: 'dolor ipsum',
          description: 'dolor ipsum',
        }],
      }],
    },
    deadline: '2020-06-27T03:00:00.043Z',
    loading: { b: false },
    message: 'tes',
    setMessageAction: jest.fn(),
  };
});

jest.mock('../../../../hooks/useTimer', () => jest.fn(() => ({
  time: ['', '', ''],
})));

describe('src/components/fragments/PaymentBankTransfer', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<PaymentBankTransfer />);
    expect(tree).toMatchSnapshot();

    useParams.mockImplementationOnce(() => ({ bank: 'vamandiri' }));
    const result = PaymentBankTransfer();
    expect(result.props.children[0].type).toBe(BankDetail);

    useParams.mockImplementationOnce(() => ({ detail: 'bank-transfer' }));
    const result1 = PaymentBankTransfer();
    expect(result1.props.children[0].type).toBe(ListBank);
  });

  test('render loading state', () => {
    useSelector.mockImplementationOnce(() => ({ loading: { b: true } }));

    const result = BankDetail();
    expect(result.props.children[1].props.children.props.className).toBe('loading');
    expect(result.props.children[2].props.children[1].props.className).toBe('loading');
  });

  test('ListBank', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<ListBank />);
    expect(tree).toMatchSnapshot();

    const dispatch = jest.fn();
    useDispatch.mockImplementationOnce(() => dispatch);

    const result = ListBank();
    result.props.children.props.children[0].props.onClick();
    expect(dispatch).toHaveBeenCalled();
  });

  test('BankDetail', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<BankDetail />);
    expect(tree).toMatchSnapshot();

    const dispatch = jest.fn();
    const select = jest.fn();
    useDispatch.mockImplementationOnce(() => dispatch);
    useRef.mockImplementationOnce(() => ({ current: { select } }));
    window.document.execCommand = jest.fn();
    const result = BankDetail();
    result.props.children[2].props.children[2].props.onClick({ target: { focus: jest.fn() } });
    expect(select).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalled();
  });

  test('ListBank no bank', () => {
    useSelector.mockImplementationOnce(() => ({ bankInfo: '' }));

    const result = ListBank();
    expect(result).toBe(null);
  });

  test('HowToPay', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<HowToPay />);
    expect(tree).toMatchSnapshot();

    // const detailBank = jest.fn();
    const setIdxOpen = jest.fn();
    useState.mockImplementationOnce(() => [1, setIdxOpen]);
    // useSelector.mockImplementationOnce(() => ({ detailBank }));
    const result = HowToPay();
    result[0].props.children[0].props.onClick();
    result[0].props.children[0].props.children[1].props.onClick();
    // expect(setIdxOpen).toHaveBeenNthCalledWith(1, 0);
    // expect(setIdxOpen).toHaveBeenNthCalledWith(2, -1);
    // expect(result[0].props.children[0].props.children[1].props.className).toBe('open');
  });

  test('Notif', () => {
    const dispatch = jest.fn();
    const setNotif = jest.fn();
    const message = jest.fn();
    useDispatch.mockImplementationOnce(() => dispatch);
    useSelector.mockImplementationOnce(fn => {
      fn({});
      return { message };
    });
    useContext.mockImplementationOnce(() => ({ setNotif }));
    Notif();
    expect(setNotif).toHaveBeenCalledWith(message);
  });

  test('Timer', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Timer />);
    expect(tree).toMatchSnapshot();

    useTimer.mockImplementationOnce(() => ({
      time: ['', '', ''],
    }));
  });
});
