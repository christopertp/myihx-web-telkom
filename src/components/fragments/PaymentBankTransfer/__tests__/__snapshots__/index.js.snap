// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/fragments/PaymentBankTransfer BankDetail 1`] = `
<section>
  <IconButton
    className=""
    name="close_red"
    onClick={[Function]}
    title=""
    to="/"
    type="button"
  />
  <header>
    <img
      alt="logo bank"
      src=""
    />
  </header>
  <section>
    <Heading
      className=""
    >
      Account Number
    </Heading>
    <input
      readOnly={true}
      type="text"
      value="396525"
    />
    <Button
      className=""
      disabled={false}
      onClick={[Function]}
      to=""
      type="button"
      variant="red"
    >
      COPY
    </Button>
  </section>
  <section>
    <Heading
      className=""
    >
      Transfer Amount
    </Heading>
    <strong>
      Rp 
      392.5
    </strong>
    <Heading
      className=""
    >
      Payment Code
    </Heading>
    <strong />
  </section>
  <Timer />
  <section>
    <Heading
      className=""
    >
      Lihat cara pembayaran
    </Heading>
    <HowToPay />
  </section>
  <Button
    className=""
    disabled={false}
    onClick={null}
    to="/"
    type="button"
    variant="red"
  >
    TUTUP
  </Button>
</section>
`;

exports[`src/components/fragments/PaymentBankTransfer HowToPay 1`] = `
Array [
  <Card
    className=""
    disabled={false}
    onClick={null}
    style={
      Object {
        "height": "3.5rem",
      }
    }
    to=""
    variant="bordered"
  >
    <header
      onClick={[Function]}
    >
      <h6>
        MOBILE
      </h6>
      <IconButton
        className=""
        name="chevron_red_down"
        onClick={[Function]}
        title=""
        to=""
        type="button"
      />
    </header>
    <ul>
      <li>
        <Bullet
          className=""
          color={
            Array [
              "#ee3124",
              "#cccccc",
            ]
          }
          connect={false}
          size="0.75rem"
        />
        <h5>
          dolor ipsum
        </h5>
        <p
          style={
            Object {
              "paddingBottom": "0.7rem",
            }
          }
        >
          dolor ipsum
        </p>
      </li>
    </ul>
  </Card>,
]
`;

exports[`src/components/fragments/PaymentBankTransfer ListBank 1`] = `
<WhiteCardPlain
  back=""
  close=""
  help={false}
  icon=""
  size="large"
  subtitle="Pilih Bank yang dituju"
  title="Pilih Bank"
>
  <section>
    <section
      onClick={[Function]}
    >
      <strong />
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </section>
  </section>
</WhiteCardPlain>
`;

exports[`src/components/fragments/PaymentBankTransfer Timer 1`] = `
<div>
  <img
    alt="time"
    src="/assets/ic_selection_time.svg"
  />
  <span>
    Complete payment by
  </span>
  <span>
    
    :
    
    :
    
  </span>
</div>
`;

exports[`src/components/fragments/PaymentBankTransfer render 1`] = `
<React.Fragment>
  <ListBank />
  <Notif />
</React.Fragment>
`;
